src/AWS/Orleans.Persistence.S3/Provider/S3GrainStorage.cs

```diff
@@ -0,0 +1,128 @@
+                        .ConfigureAwait(false);
+
+                    this.serializationManager.DeserializeToState(grainState, data);
+                }
+            }
+            catch (AmazonS3Exception e) when (e.StatusCode == HttpStatusCode.NotFound)
+            {
+                grainState.State = grainState.CreateDefaultState();
+            }
+            catch (Exception e)
+            {
+                this.logger.LogError(new EventId((int)ErrorCode.StorageProviderBase), e,
+                    "Error Reading: GrainType={GrainType} GrainId={GrainId} ETag={ETag} to BucketName={BucketName} Exception={Exception}",
+                    grainType, grainReference, grainState.ETag, this.options.BucketName, e.Message);
+                throw;
+            }
+        }
+
+        public async Task WriteStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+        {
+            if (this.logger.IsEnabled(LogLevel.Trace))
+            {
+                this.logger.Trace(ErrorCode.StorageProviderBase,
+                    "Writing: GrainType={GrainType} GrainId={GrainId} from Bucket={BucketName}",
+                    grainType, grainReference, this.options.BucketName);
+            }
+
+            try
+            {
+                await this.s3.PutObjectAsync(new PutObjectRequest
+                {
+                    BucketName = this.options.BucketName,
+                    Key = this.keyFormatter.FormatKey(this.name, grainType, grainReference),
+                    InputStream = this.serializationManager
+                        .SerializeFromState(grainState)
+                        .AsMemoryStream()
+                }).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                this.logger.LogError(new EventId((int)ErrorCode.StorageProviderBase), e,
+                    "Error Writing: GrainType={GrainType} GrainId={GrainId} ETag={ETag} to BucketName={BucketName} Exception={Exception}",
+                    grainType, grainReference, grainState.ETag, this.options.BucketName, e.Message);
+                throw;
+            }
+        }
+
+        public async Task ClearStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+        {
+            try
+            {
+                await this.s3.DeleteObjectAsync(new DeleteObjectRequest
+                {
+                    BucketName = this.options.BucketName,
+                    Key = this.keyFormatter.FormatKey(this.name, grainType, grainReference)
+                }).ConfigureAwait(false);
+            }
+            catch (AmazonS3Exception e) when (e.StatusCode == HttpStatusCode.NotFound)
+            {
+                // Ignore not found errors
+            }
+        }
+
+        public void Participate(ISiloLifecycle lifecycle) => lifecycle.Subscribe<S3GrainStorage>(this.options.InitStage, this);
```

Comment:
Will probably remove `ILifecycleParticipant` support as it's not really needed any more